src/sys/windows/selector.rs

```diff
@@ -1,596 +1,545 @@
+        &self.cp
+    }
+
+    fn update_sockets_events(&self) -> io::Result<()> {
+        {
+            let mut update_queue = self.update_queue.lock().unwrap();
+            loop {
+                let sock = match update_queue.pop_front() {
+                    Some(sock) => sock,
+                    None => break,
+                };
+                let ptr = Arc::into_raw(sock.clone());
+                let mut sock_internal = sock.lock().unwrap();
+                if !sock_internal.is_pending_deletion() {
+                    sock_internal.update(ptr).unwrap();
+                }
+            }
+        }
+        self._release_unused_afd();
+        Ok(())
     }
 
-    pub fn capacity(&self) -> usize {
-        self.events.capacity()
+    fn update_sockets_events_if_polling(&self) -> io::Result<()> {
+        if self.active_poll_count.load(Ordering::SeqCst) > 0 {
+            return self.update_sockets_events();
+        }
+        Ok(())
     }
 
-    pub fn get(&self, idx: usize) -> Option<&Event> {
-        self.events.get(idx)
+    fn feed_events(&self, events: &mut Events, iocp_events: &[CompletionStatus]) {
+        {
+            let mut update_queue = self.update_queue.lock().unwrap();
+            for iocp_event in iocp_events.iter() {
+                if iocp_event.overlapped() as usize == 0 {
+                    events.push(Event {
+                        flags: AFD_POLL_RECEIVE,
+                        data: iocp_event.token() as u64,
+                    });
+                    continue;
+                }
+                let sock: Arc<Mutex<SockState>> =
+                    unsafe { Arc::from_raw(mem::transmute(iocp_event.overlapped())) };
+                let mut sock_guard = sock.lock().unwrap();
+                match sock_guard.feed_event() {
+                    Some(e) => {
+                        events.push(e);
+                    }
+                    None => {}
+                }
+                if !sock_guard.is_pending_deletion() {
+                    update_queue.push_back(sock.clone());
+                }
+            }
+        }
+        self._release_unused_afd();
+    }
+
+    fn _acquire_afd(&self) -> io::Result<Arc<Afd>> {
+        let mut need_alloc = false;
+        {
+            let afd_group = self.afd_group.lock().unwrap();
+            if afd_group.len() == 0 {
+                need_alloc = true;
+            } else {
+                // + 1 reference in Vec
+                if Arc::strong_count(afd_group.last().unwrap()) >= POLL_GROUP__MAX_GROUP_SIZE + 1 {
+                    need_alloc = true;
+                }
+            }
+        }
+        if need_alloc {
+            self._alloc_afd_group()?;
+        }
+        match self.afd_group.lock().unwrap().last() {
+            Some(rc) => Ok(rc.clone()),
+            None => unreachable!(),
+        }
     }
 
-    pub fn push_event(&mut self, event: Event) {
-        self.events.push(event);
+    fn _release_unused_afd(&self) {
```

Comment:
@piscisaureus is this the strategy you used in `wepoll`?

Reply:
This is better, in wepoll I never close afd driver handles.
(That said, you would want to avoid repeatedly closing one and then immediately creating a new one, which I didn't check)

Reply:
The afd poll size is same as your wepoll, so the behavior should be the same.